virttest/utils_backup.py

```diff
@@ -0,0 +1,285 @@
+        nbd_tcp_port = nbd_params.get("nbd_tcp_port", "10809")
+        nbd_type = "inet"
+        nbd_server = "server.host=%s,server.port=%s" % (nbd_host, nbd_tcp_port)
+        nbd_image_path = "nbd://%s:%s/%s" % (nbd_host, nbd_tcp_port, nbd_export)
+    elif nbd_protocol == "unix":
+        nbd_type = "unix"
+        nbd_socket = nbd_params.get("nbd_socket", "/tmp/pull_backup.socket")
+        nbd_server = "server.path=%s" % nbd_socket
+        nbd_image_path = "nbd+unix:///%s?socket=%s" % (nbd_export, nbd_socket)
+    map_from %= (nbd_export, nbd_type, nbd_server, bitmap_name)
+    # Rebase backup file to nbd image
+    rebase_cmd = "qemu-img rebase -u -f qcow2 -F raw -b %s %s"
+    rebase_cmd %= (nbd_image_path, target_path)
+    process.run(rebase_cmd, shell=True)
+    # Remove backing chain of the backup file
+    qemu_map_cmd = "qemu-img map --output=json %s -U" % map_from
+    result = process.run(qemu_map_cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    for entry in json_data:
+        if not entry["data"]:
+            qemu_io_cmd = "qemu-io -C -c \"r %s %s\" -f qcow2 %s"
+            qemu_io_cmd %= (entry["start"], entry["length"], target_path)
+            process.run(qemu_io_cmd, shell=True)
+    # remove the backing file info for the target image
+    rebase_cmd = "qemu-img rebase -u -f qcow2 -b '' %s" % target_path
+    process.run(rebase_cmd, shell=True)
+
+
+def pull_full_backup_to_file(nbd_export, target_path):
+    """
+    Dump pull-mode full backup data to a file
+
+    :param nbd_export: The nbd export info of the backup job
+    :param target_path: The path of the file to dump the data
+    """
+    cmd = "qemu-img convert -f raw %s -O qcow2 %s" % (nbd_export,
+                                                      target_path)
+    logging.debug("Execute command %s: %s", cmd,
+                  process.run(cmd, shell=True).stdout_text)
+
+
+def get_img_data_size(img_path, driver='qcow2'):
+    """
+    Get the actual data size of a image
+
+    :param img_path: The path to the image
+    :return: The actual data size of the image in 'MB'.
+    """
+    cmd = "qemu-img map -f %s %s --output=json -U" % (driver, img_path)
+    result = process.run(cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    summary = 0
+    for entry in json_data:
+        if entry["data"]:
+            summary += entry["length"]
+    return summary/1024/1024
+
+
+def get_img_data_map(img_path, driver='qcow2'):
+    """
+    Get the data map of a image
+
+    :prama img_path: The path to the image
+    :return: The data map of the image
+    """
+    cmd = "qemu-img map -f %s %s --output=json -U" % (driver, img_path)
+    result = process.run(cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    data_map = []
+    for entry in json_data:
+        if entry["data"]:
+            data_map.append(entry)
+    return data_map
+
+
+def dd_data_to_vm_disk(vm, disk, bs, seek, count):
```

Comment:
Would you please move it to other files, it's not very close to backup feature I think, how do you think?

Reply:
will move to utils_disk.py